protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,100 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	"github.com/nats-io/nats.go"
+)
+
+// Protocol is a reference implementation for using the CloudEvents binding
+// integration. Protocol acts as both a NATS client and a NATS handler.
+type Protocol struct {
+	conn *nats.Conn
+
+	Consumer *consumer
+
+	Sender *sender
+
+	connOwned bool // whether this protocol created the nats connection
+}
+
+// NewProtocol creates a new NATS protocol.
+func NewProtocol(ctx context.Context, url, sendSubject string, natsOpts []nats.Option, opts ...ProtocolOption) (*Protocol, error) {
+	conn, err := nats.Connect(url, natsOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewProtocolFromConn(ctx, conn, sendSubject, opts...)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	p.connOwned = true
+
+	return p, nil
+}
+
+// NewProtocolFromConn creates a new NATS protocol with the given nats connection.
+func NewProtocolFromConn(ctx context.Context, conn *nats.Conn, sendSubject string, opts ...ProtocolOption) (*Protocol, error) {
```

Comment:
question: should we accept an interface for nats.Conn here to simplify mocking or hardcode it to this type?

Reply:
unfourtunately, the jetstream library explictly uses a "pointer to struct"